src/sync/partial_sync.cpp

```diff
@@ -56,6 +83,153 @@ void update_schema(Group& group, Property matches_property)
         ObjectStore::apply_additive_changes(group, required_changes, true);
 }
 
+bool validate_existing_subscription(std::shared_ptr<Realm> realm,
+                                    std::string const& name,
+                                    std::string const& query,
+                                    std::string const& matches_property,
+                                    ObjectSchema const& result_sets_schema)
+{
+    TableRef table = ObjectStore::table_for_object_type(realm->read_group(), result_sets_type_name);
+    size_t name_idx = result_sets_schema.property_for_name("name")->table_column;
+    auto existing_row_ndx = table->find_first_string(name_idx, name);
+    if (existing_row_ndx == npos)
+        return false;
+
+    Object existing_object(realm, result_sets_schema, table->get(existing_row_ndx));
+
+    CppContext context;
+    std::string existing_query = any_cast<std::string>(existing_object.get_property_value<util::Any>(context, "query"));
+    if (existing_query != query)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different query.");
+
+    std::string existing_matches_property = any_cast<std::string>(existing_object.get_property_value<util::Any>(context, "matches_property"));
+    if (existing_matches_property != matches_property)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different result type.");
+
+    return true;
+}
+
+class WorkQueue {
+public:
+    ~WorkQueue()
+    {
+        reset();
+    }
+
+    void reset()
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = true;
+        }
+        m_cv.notify_one();
+
+        if (m_thread.joinable())
+            m_thread.join();
+
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_stopping = false;
+        }
+    }
+
+    void enqueue(std::function<void()> function)
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_mutex);
+            m_queue.push_back(std::move(function));
+
+            if (m_stopped)
+                create_thread();
+        }
+        m_cv.notify_one();
+    }
+
+private:
+    void create_thread()
+    {
+        if (m_thread.joinable())
+            m_thread.join();
+
+        m_thread = std::thread([this] {
+            std::vector<std::function<void()>> queue;
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            while (!m_stopping &&
```

Comment:
I think this first m_stopping check is redundant.

Reply:
Why's that? Without it the loop will only exit if `wait_for` returns false due to none of the conditions being met before the timeout expires.

Reply:
I missed that the behavior is different for m_stopping being true in the first check and in the predicate (and that difference is correct).